### PR TITLE
Fix pointer type

### PR DESCRIPTION
**Pull request by kroening**
Pointers in the frontend (before type checking) are different, and don't yet obey the type invariant of pointer_typet.

### CODE REVIEW

File: src/ansi-c/c_typecheck_type.cpp [hunk 1]
```diff
@@ -1492,19 +1492,22 @@ void c_typecheck_baset::adjust_function_parameter(typet &type) const
 {
   if(type.id()==ID_array)
   {
+    source_locationt source_location=type.source_location();
```
Comment (reuk) [inline src/ansi-c/c_typecheck_type.cpp:1495]: Should be const.

File: src/ansi-c/c_typecheck_type.cpp [hunk 2]
```diff
@@ -1492,19 +1492,22 @@ void c_typecheck_baset::adjust_function_parameter(typet &type) const
...
     type=pointer_type(type.subtype());
-    type.remove(ID_size);
-    type.remove(ID_C_constant);
+    type.add_source_location()=source_location;
   }
   else if(type.id()==ID_code)
   {
     // see ISO/IEC 9899:1999 page 199 clause 8,
     // may be hidden in typedef
+    source_locationt source_location=type.source_location();
```
Comment (reuk) [inline src/ansi-c/c_typecheck_type.cpp:1503]: Should be const.

**Review by reuk — COMMENTED** [review]
We have no way of enforcing, or even *seeing* that the pointer ends up with a width in all cases. At least if we use the `pointer_typet` constructor, we know that the object is a valid instance of that type. I don't understand how this change is an improvement.

**Comment by kroening** [comment]
The type invariant problem is now addressed. There are now frontend_pointers, which do not have a width, which then get turned into pointers, which do have it.

File: src/ansi-c/ansi_c_convert_type.cpp [hunk 3]
```diff
@@ -216,13 +216,20 @@ void ansi_c_convert_typet::read_rec(const typet &type)
...
   {
-    // pointers have a width, much like integers
-    pointer_typet tmp=to_pointer_type(type);
-    tmp.set_width(config.ansi_c.pointer_width);
+    // We turn ID_frontend_pointer to ID_pointer
+    // Pointers have a width, much like integers,
+    // which is added here.
+    typet tmp(type);
+    tmp.id(ID_pointer);
+    tmp.set(ID_width, config.ansi_c.pointer_width);
```
Comment (reuk) [inline src/ansi-c/ansi_c_convert_type.cpp:226]: This should use the proper pointer_type constructor

**Review by reuk — CHANGES_REQUESTED** [review]

**Review by peterschrammel — APPROVED** [review]
lgtm modulo addressing @reuk's comments

**Pull request merged**